common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java

```diff
@@ -243,6 +247,104 @@ public static void load(String name, ClassLoader loader) {
         }
     }
 
+    /**
+     * Loading the native library through the specified ClassLoader.
+     * @param loader - The specified ClassLoader where the native library will be loaded into
+     * @param name - The native library path or name
+     * @param absolute - Whether the native library will be loaded by path or not
+     */
+    private static void loadLibrary(final ClassLoader loader, final String name, final boolean absolute) {
+        try {
+            // Make sure the helper is belong to the target ClassLoader.
+            final Class<?> newHelper = tryToLoadClass(loader, NativeLibraryUtil.class);
+            AccessController.doPrivileged(new PrivilegedAction<Object>() {
+                @Override
+                public Object run() {
+                    try {
+                        /*
+                         * Invoke the helper to load the native library, if succeed, then the native
+                         * library belong to the specified ClassLoader.
+                         */
+                        Method method = newHelper.getMethod("loadLibrary",
+                                new Class<?>[] { String.class, boolean.class });
+                        return method.invoke(null, name, absolute);
+                    } catch (Exception e) {
+                        throw new IllegalArgumentException("Load library failed!", e);
+                    }
+                }
+            });
+        } catch (Exception e) { // Should by pass the UnsatisfiedLinkError here!
+            logger.debug("Unable to load the library: " + name + '.', e);
+            NativeLibraryUtil.loadLibrary(name, absolute);
+        }
+    }
+
+    /**
+     * Try to load the helper Class into specified ClassLoader.
+     * @param loader - The specified ClassLoader
+     * @param helper - The helper Class
+     * @return A new helper Class defined in the specified ClassLoader.
+     * @throws ClassNotFoundException Helper class not found or loading failed
+     */
+    private static Class<?> tryToLoadClass(final ClassLoader loader, final Class<?> helper)
+            throws ClassNotFoundException {
+        try {
+            return loader.loadClass(helper.getName());
+        } catch (ClassNotFoundException e) {
+            // The helper class NOT found in target ClassLoader, we have to define the helper class.
+            final byte[] classBinary = classToByteArray(helper);
+            return AccessController.doPrivileged(new PrivilegedAction<Class<?>>() {
+                @Override
+                public Class<?> run() {
+                    try {
+                        // Define the helper class in the target ClassLoader,
+                        //  then we can call the helper to load the native library.
+                        Method defineClass = ClassLoader.class.getDeclaredMethod("defineClass", String.class,
```

Comment:
To my view, the answer is not possible. Because the `NativeLibraryLoader` is a utility. Its method such as `load(String, ClassLoader)` is invoked with a **target** `ClassLoader` and we need to define the helper class into it, that is a dynamic process. For example, the `netty-tcnative` library is loaded by `OpenSsl` in `netty-handler` with the `SSL`'s `ClassLoader`, we have to inject the helper class into `SSL`'s `ClassLoader`. But for `netty-transport-native-epoll` library is loaded by `Native` with its `ClassLoader`, then we have to inject the helper class into `Native`'s `ClassLoader`.
The helper class injection process is nothing to do with `NativeLibraryLoader` itself . I have no idea how to make it static.

BTW, there are two native library involved in netty, both have the same OSGi loading issue. And there has another solution to this issue:
- `netty-tcnative` case, which provided a loading helper class `org.apache.tomcat.jni.Library`. Just like 9f2a2135, we should modify `OpenSsl`, instead of calling the `NativeLibraryLoader`, trying the helper class first (Depends on a properly `Bundle-NativeCode` manifest entry). _The helper class `Library` in `netty-tcnative` does NOT support loading library by absolute path._
- `netty-transport-native-epoll` case, we should modify `Native`, instead of calling the `NativeLibraryLoader`, trying the `System.load()` first (Depends on a properly `Bundle-NativeCode` manifest entry).
This solution is more simple, but there has a drawback: the library loading process is diff between OSGi and non-OSGi. In the OSGi, the library in `$PATH` has a higher priority than the one in the bundle. The non-OSGi is just the reverse. That's why I submitted a more complicated solution:)
Thanks for all of the comments!